### PR TITLE
l0-setup dockercfg flag update

### PR DESCRIPTION
**Pull request by jonathanstallings**
Addresses Issue #11.

Principal changes include:
- Add `--dockercfg` flag on `l0-setup apply` allowing passing a path to valid`dockercfg` file. This file will be copied to the instance directory.
- Reword user prompt displayed when no `dockercfg` file supplied by user.
- Create default `dockercfg` file before showing user prompt.
- Add simple JSON validation of `dockercfg` file.
- Begin refactoring for easier testability.
- Add simple unit test for `validateDockercfg`.

Further work would be needed to add more test coverage.

### CODE REVIEW

File: setup/context/commands.go [hunk 1]
```diff
@@ -5,14 +5,15 @@ import (
...
-	"github.com/quintilesims/layer0/common/aws/provider"
-	"github.com/quintilesims/layer0/common/aws/s3"
-	"github.com/quintilesims/layer0/common/config"
 	"io/ioutil"
 	"net/http"
 	"os"
 	"strings"
 	"time"
+
+	"github.com/quintilesims/layer0/common/aws/provider"
```
Comment (diemonster) [inline setup/context/commands.go]: [nitpick] You're probably using `goimports` rather than `gofmt`, which creates this line break between stdlib packages and 3rd party libraries. Our styleguide thus far has been to use `gofmt`, which does not add this line break.

Comment (jonathanstallings) [inline setup/context/commands.go]: Right. I remember you mentioned that. Could you send me a link to the styleguide?

Comment (jonathanstallings) [inline setup/context/commands.go]: Changed to using `gofmt` instead of `goimports` from now on. Thanks for the reminder!

File: setup/context/commands_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"os"
+	"testing"
+)
+
+type MockFileIO struct {
+	data map[string][]byte
+}
+
+func (mock MockFileIO) ReadFile(path string) ([]byte, error) {
```
Comment (diemonster) [inline setup/context/commands_test.go]: I would recommend making your method receivers into pointers. This is a pretty good article on why that can be important, aside from consistency with the majority of our other methods: https://nathanleclaire.com/blog/2014/08/09/dont-get-bitten-by-pointer-vs-non-pointer-method-receivers-in-golang/

Comment (jonathanstallings) [inline setup/context/commands_test.go]: Thanks! I'll check that out.

File: setup/context/commands.go [hunk 3]
```diff
@@ -393,35 +430,60 @@ func Endpoint(c *Context, syntax string, insecure, dev, quiet bool) error {
...
+func checkDockercfg(c *Context, force bool, io FileIO) error {
 	path := fmt.Sprintf("%s/dockercfg", c.InstanceDir)
 
-	// if file doesn't exist
-	if _, err := os.Stat(path); err != nil {
+	// if file does not exist
+	if _, err := io.Stat(path); err != nil {
+
+		// write empty dockercfg file
+		if err := io.WriteFile(path, []byte("{}"), 0666); err != nil {
```
Comment (diemonster) [inline setup/context/commands.go]: [nit] the file permission here is `0666` --- I'd recommend changing it to `0660`, not sure why `everyone` would need to have read/write access to an auth file, esp on a multi-user system (this was like this prior to your PR, btw, missed it the first time!)

File: setup/context/commands.go (hunk 1 above)
Comment (diemonster) [inline setup/context/commands.go]: There is no styleguide aside from the code b/c `gofmt` takes care of guiding our style 🎩

File: setup/context/commands.go [hunk 4]
```diff
@@ -63,8 +87,21 @@ func Apply(c *Context, force bool) error {
 		}
 	}
 
-	// write empty dockercfg file if it doesn't exist
-	if err := checkDockercfg(c, force); err != nil {
+	var dockercfg string
+	dockercfg_flag, ok := c.Flags["dockercfg"]
```
Comment (zpatrick) [inline setup/context/commands.go]: `dockercfg` should not be a context flag, it should be passed in as a param to `Apply` just like `force` flag is (will go over more in main)

File: setup/context/commands.go [hunk 5]
```diff
@@ -63,8 +87,21 @@ func Apply(c *Context, force bool) error {
...
+
+	if ok && dockercfg_flag != nil && *dockercfg_flag != "" {
+		fmt.Printf("Detected dockercfg: `%s`\n", *dockercfg_flag)
+		dockercfg = *dockercfg_flag
```
Comment (zpatrick) [inline setup/context/commands.go]: go uses camel case, not underscore: `dockercfgFlag`

File: setup/context/commands.go [hunk 6]
```diff
@@ -393,35 +430,60 @@ func Endpoint(c *Context, syntax string, insecure, dev, quiet bool) error {
...
 			}
 		}
 
-		// write empty dockercfg file
-		if err := ioutil.WriteFile(path, []byte("{}"), 0666); err != nil {
-			return fmt.Errorf("Failed to write 'dockercfg': %v", err)
-		}
+	}
+
+	if err := validateDockercfg(path, io); err != nil {
```
Comment (zpatrick) [inline setup/context/commands.go]: [nit] shorten this block to: `return validateDockercfg(path, io)`

File: setup/context/commands_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
+package context
```
Comment (zpatrick) [inline setup/context/commands_test.go]: I'm pretty hesitant to give the :+1:  on these unit tests and introducing the whole `FileIO` interface for a couple of reasons:

* This tool isn't setup to be unit tested. In my opinion, adding more interfaces, classes, and mocks on top of the current architecture isn't a step in the right direction - it just adds more complexity and maintenance costs to an already complex tool. If we want unit tests, we should do it right or not at all. 
* Creating your own mock class (`MockIO`) is a pit we fell into when first starting go. The "better" way of generating mocks is is via a tool called [gomock](github.com/golang/mock). That being said, regardless of how the mock class is generated/updated, the larger issue I have with it is the additional maintenance cost. 
* This is echoing the first point a bit - but these unit tests don't really _add_ anything. They essentially just check if `json.Marshal` works. If having unit tests don't give you confidence in your code - why have them at all? Don't get me wrong - I'm not trying to say that Jonathan writes bad unit tests - what I **am** saying is that the current architecture is so bad that it's damn near impossible for anyone to write a useful unit test. 

TLDR: I think it would be better to drop the unit tests and associated interfaces/mocks/etc from this PR in favor of re-architecting this tool to be more unit testable.

File: setup/main.go [hunk 8]
```diff
@@ -74,6 +75,10 @@ func main() {
 
 	app.Command("apply", "Create/Update a Layer0", func(cmd *cli.Cmd) {
```
Comment (zpatrick) [inline setup/main.go]: `dockercfg` is not a flag for the `context` object, but for the `apply` command. This should look like:
```
app.Command("apply", "Create/Update a Layer0", func(cmd *cli.Cmd) {
		flags := loadFlags(cmd, []string{"access_key", "secret_key", "region"})
		force := cmd.BoolOpt("force", false, "Set this flag to skip prompting on a missing dockercfg file")
		dockercfg := cmd.StringOpt("dockercfg", "", "Path to valid dockercfg file")

		vpc := cmd.StringOpt("vpc", "", "VPC id to target.  Will create new VPC if blank.")
		flags["vpc_id"] = vpc

		command := func(c *context.Context) error {
			return context.Apply(c, *force, *dockercfg)
		}

		flagsCommand(cmd, command, flags)
	})
```

File: setup/context/commands_test.go (hunk 7 above)
Comment (diemonster) [inline setup/context/commands_test.go]: I had a feeling you would say as much about `gomock`. I think @jonathanstallings work here is valuable in learning _why_ we use Gomock, and in that sense, this was not a pointless exercise. As for the value of the tests themselves and the "all or nothing" approach, I'm inclined to agree with you.

Devil's advocate argument re: `MockIO` is that for new developers coming into this project, the need for Gomock is not entirely obvious. I'd like us to keep that in mind when people attempt to add tests in the future (if they even try to add tests!)

File: setup/context/commands.go (hunk 5 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: Of course. Should I change the case on `vpc_flag` etc. from which I modeled this from as well?
https://github.com/jonathanstallings/layer0/blob/c8d151db6033d76ac354821be1940df9cab087e3/setup/context/commands.go#L60

File: setup/context/commands.go (hunk 6 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: Yup!

File: setup/context/commands.go (hunk 4 above)
Comment (jonathanstallings) [inline setup/context/commands.go]: Got it. Thanks, zpatrick!

File: setup/main.go (hunk 8 above)
Comment (jonathanstallings) [inline setup/main.go]: Thanks! I have made this change.

File: setup/context/commands_test.go (hunk 7 above)
Comment (jonathanstallings) [inline setup/context/commands_test.go]: Thanks, both. I appreciate the discussion here. Part of why I opted to create the mock class directly was because of my unfamiliarity with gomock, for sure. It also served as a bit of a learning exercise for working with Go. I will certainly look into using the best tools for the project.

I have to differ with the percieved value of the unit tests. Although it is true that the _current_ implementation is essentially just wrapping `json.Marshal`, that could certainly change; e.g., more robust validation against certain schemas could be added or it could be tweaked another way. The unit tests are there as sentinels to ensure that behavior works as expected when things are changed.

Regardless, I agree with the larger points expressed above. The unit tests have been dropped, and the mock class has been removed.

**Comment by jonathanstallings** [comment]
Thank you, @diemonster and @zpatrick for your thorough and useful feedback. I understand that the `l0-setup` tool may be re-architected, and thus this PR may not have relevance long term.

In any case, it was a great way for me to learn a bit more about Go and layer0, which is a project I would like to support in the future.
